### DISCUSSION TITLE
TCP socket

### DISCUSSION THREAD

**Discussion by lestat0712** [polarfire-soc/polarfire-soc-documentation#110] [Bare metal embedded software] [answered]
Hi everyone, 

I have the icicle kit and would like to implement a client tcp socket. I have tried modifying the mpfs-mac-mcc-stack to do this, for both client and server socket but with no luck. Wonder has anyone implemented this using the icicle kit and one of the bare metal examples? any suggestions ? 

Thanks very much in advance and best regards

**Comment by rockoguido** [comment]
Hi..

I recently did a client server app for a PolarFire SOC customer, on the SOC side the server ran and the client was on a Linux desktop.   Of course it could be reversed with the client on the SOC.  It’s a socket based app

Is that of interest?

Regards,

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 9890 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

[Text  Description automatically generated]
<http://www.microchip.com/emasters>Visit our new learning site at www.microchip.com/mu<http://www.microchip.com/mu>

From: lestat0712 ***@***.***>
Sent: Thursday, June 30, 2022 8:57 AM
To: polarfire-soc/polarfire-soc-documentation ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [PolarFire-SoC] TCP socket (Discussion #110)

EXTERNAL EMAIL: Do not click links or open attachments unless you know the content is safe

Hi everyone,

I have the icicle kit and would like to implement a client tcp socket. I have tried modifying the mpfs-mac-mcc-stack to do this, for both client and server socket but with no luck. Wonder has anyone implemented this using the icicle kit and one of the bare metal examples? any suggestions ?

Thanks very much in advance and best regards

**Reply by lestat0712** [reply]
Hi Dave, 

Thanks for your answer. Yes that would be of great interest ! to be honest I have flexibility as to where to run the server and the client. I am new to the Polarfire so I am still learning about it. 
As I said, I tried modifying the mpfs-mac-mcc-stack as a starting point to implement both, TCP client and server sockets. When I tried the server socket, socket init, binding and listening seemed to work but then when I tried to connect from the client side , it timed out. the polarfire was receiving packets though as they were being logged, but the FW was not entering  the TCP_FiniteStateMachine () with the incoming packets/ client connection attempts. Socket status was in progress, port and IP addresses looked correct...... and when I tried to implement the client, it was timing out after sending the SYN packet ......

As I said, I am new to the Polarfire SoC FPGA and the baremetal examples, so I am sure I am missing a lot of things. Thanks very much. 

Kind regards,
Javier

**Comment by rockoguido** [comment]
Hi Javier,

Ok, a few attachments have been included…


  1.  SoftConsole Project for the target.   One important note on this project.   You need to build with a different set of tools than what SoftConsole uses.   In another words you cannot use the default riscv64-unknown-elf-gcc cross compiler since we are including networking based code.   I used the cross tools that are built via Buildroot (riscv64-unknown-linux-gnu-gcc).   See the Project Settings -> Toolchains setting
     *   Naturally you will be building this project via SoftConsole on a Linux host
  2.  I also attached client.c, which by default runs on the Linux host, you can just use gcc to build this.
  3.  You will want to execute the server first, of course, and then the client.

If you have any questions on this please let me know.

Regards,

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 9890 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

[Text  Description automatically generated]
<http://www.microchip.com/emasters>Visit our new learning site at www.microchip.com/mu<http://www.microchip.com/mu>

**Reply by AntonKrug** [reply]
Hi @rockoguido,

Looks like you sent him a socket implementation running on Linux, while it looks to me @lestat0712 is looking for bare metal solution.

**Comment by rockoguido** [comment]
ok adding archives since they did not get attached originally
[client.zip](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9029318/client.zip)
[socket_server.zip](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9029321/socket_server.zip)

The code inserted above is the client, same as client.zip here

**Reply by lestat0712** [reply]
Thanks a million @rockoguido for your support. @AntonKrug is right, I was looking for a bare metal solution (yours seems to be for a Linux). Nevertheless, thanks  so much for your time and for sharing the code! I really appreciate it

**Comment by petermcs-01** [comment, marked as answer]
The mpfs-mac-mcc-stack is very much work in progress with much of the low level stuff working and a UDP example based on BACnet. I had hoped to complete the FTP client example to enable me test and debug the TCP/IP side of things but haven't gotten to that yet. Your best bet for a more complete TCP/IP stack would be to look at the LwIP and FreeRTOS example.

**Reply by lestat0712** [reply]
Thanks @petermcs-01 for your answer and support, I will have a look at those examples.

**Reply by AntonKrug** [reply]
@petermcs-01, would be worthy for @lestat0712 to look at our Zephyr port as well? As that is more feature rich than FreeRTOS.

**Reply by petermcs-01** [reply]
@AntonKrug Unfortunately support for the GEM on Zephyr is not implemented yet so FreeRTOS/LwIP is the best option for now.

**Comment by rockoguido** [comment]
Hi Anton,

If I look at Javier’s original message he seems to be looking for a client TCP socket application and server too.   However you’re right he’s talking about bare metal too.
Yes to me if I’m doing sockets I have an underlying OS/networking stack so at the end of the day it may not be of much use to him.

I’m puzzled by the combo of bare metal and TCP based sockets.    Well, I just had this app that I did for a customer so thought it might be of some value.

Thx!

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 9890 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

[Text  Description automatically generated]
<http://www.microchip.com/emasters>Visit our new learning site at www.microchip.com/mu<http://www.microchip.com/mu>
